subprojects/core/src/main/java/org/gradle/execution/plan/ExecutionNodeAccessHierarchy.java

```diff
@@ -69,13 +69,30 @@ public void visitChildren(Iterable<NodeAccess> values, Supplier<String> relative
             @Override
             public void visitChildren(Iterable<NodeAccess> values, Supplier<String> relativePathSupplier) {
                 String relativePathFromLocation = relativePathSupplier.get();
-                if (filter.isSatisfiedBy(new ReadOnlyFileTreeElement(new File(location + "/" + relativePathFromLocation), relativePathFromLocation, stat))) {
+                if (relativePathMatchesSpec(filter, new File(location, relativePathFromLocation), relativePathFromLocation)) {
                     values.forEach(this::addNode);
                 }
             }
         });
     }
 
+    private boolean relativePathMatchesSpec(Spec<FileTreeElement> filter, File element, String relativePathString) {
+        boolean elementIsFile = element.isFile();
+        RelativePath relativePath = RelativePath.parse(elementIsFile, relativePathString);
+        if (!filter.isSatisfiedBy(new ReadOnlyFileTreeElement(element, relativePath, stat))) {
+            return false;
+        }
+        // All parent paths need to match the spec as well, since this is how we implement the file system walking for file tree.
+        File parentFile = element.getParentFile();
+        RelativePath parentRelativePath = relativePath.getParent();
```

Comment:
I'd love to fix the nullability of `getParent()` (aka it can return null), though it seems like this can break Kotlin build scripts: https://github.com/gradle/gradle/pull/8975

Can we do something like changing the nullability for a major release?

Reply:
Maybe instead of changing the nullability we could introduce a new method and deprecate this one?

Reply:
I talked to Paul about this. There is an `@UnderMigration` annotation which we may be able to use. Though that will require some changes in the Kotlin DSL, especially how we compile build scripts.